File: SatelliteClasses/SatelliteClasses/Geometry/Ellipse.py
import speedmath
from Geometry.ConicSection import ConicSection

class Ellipse(ConicSection):

    '''
    This class discrabes ellipse. More information: https://en.wikipedia.org/wiki/Ellipse
    Class Ellipse inherits from class ConicSection.

    Protected attributes:
        1) _Eccentricity
        2) _SemiMajorAxis

    Properties:
        1) Eccentricity
        2) SemiMajorAxis
        3) SemiMinorAxis
        4) AspectRatio
        5) PerifocusDistance
        6) ApothecureDistance
        7) SemiLatusRectum
        8) LinearEccentricity
        9) FocalParameter

    Setters:
        1) Eccentricity
        2) SemiMajorAxis
    '''
    
    __slots__ = ("_Eccentricity", "_SemiMajorAxis")

    def __init__(self):
        '''
        This constructor sets:
        _Eccentricity = 0
        _SemiMajorAxis = 6400000
        '''
        ConicSection.__init__(self)
        pass

    @property
    def SemiMinorAxis(self):
        '''
        Returns semi-minor axis.
        '''
        return speedmath.EllipseSemiminorAxis(self._Eccentricity, self._SemiMajorAxis)

    @property
    def AspectRatio(self):
        '''
        Returns aspect ratio.
        '''
        return speedmath.EllipseAspectRatio(self._Eccentricity)

    @property
    def PerifocusDistance(self):
        '''
        Returns perifocus distance.
        '''
        return speedmath.EllipsePerifocusDistance(self._Eccentricity, self._SemiMajorAxis)

    @property
    def ApothecureDistance(self):
        '''
        Returns aputhecure distance.
        '''
        return speedmath.EllipseApothecureDistance(self._Eccentricity, self._SemiMajorAxis)

    @property
    def SemiLatusRectum(self):
        '''
        Returns semi-latus rectum.
        '''
        return speedmath.EllipseSemiLatusRectum(self._Eccentricity, self._SemiMajorAxis)

    @property
    def LinearEccentricity(self):
        '''
        Returns linear eccentricity.
        '''
        return speedmath.EllipseLinearEccentricity(self._Eccentricity, self._SemiMajorAxis)

    @property
    def FocalParameter(self):
        '''
        Returns focal parameter.
        '''
        return speedmath.EllipseFocalParameter(self._Eccentricity, self._SemiMajorAxis)
    
    @ConicSection.Eccentricity.setter
    def Eccentricity(self, value : float):
        '''
        Sets eccentricity. It must be a non-negative number less then 1.
        '''
        if 0.0 <= value < 1.0:
            self._Eccentricity = value
            pass
        else:
            raise ValueError("Eccentricity should be a number less than 1 and greater than or equal to 0")
        pass
    pass
    

File: SatelliteClasses/Extention/ellipse.cpp
#include "ellipse.h"
#include "usefullthings.h"
#include <cmath>

// This module contains description of the methods of the classes ConicSection and Ellipse

Ellipse::Ellipse() : ConicSection()
{

}

Ellipse::Ellipse(double eccentricity, double semi_major_axis) : 
	ConicSection(eccentricity, semi_major_axis)
{

}

Ellipse::Ellipse(PyObject* tuple) : ConicSection(tuple)
{

}

Ellipse::Ellipse(PyObject* eccentricity, PyObject* semi_major_axis) : 
	ConicSection(eccentricity, semi_major_axis)
{

}

void Ellipse::set_eccentricity(double value)
{
	if (value >= 0 && value < 1)
		eccentricity = value;
	else
		throw "Eccentricity must be a non-negative number less then 1";
}

void Ellipse::pytype_set_eccentricity(PyObject* pyvalue)
{
	double value = PyFloat_AsDouble(pyvalue);
	if (value >= 0 && value < 1)
		eccentricity = value;
	else
		throw "Eccentricity must be a non-negative number less then 1";
}

void Ellipse::set_tuple(PyObject* tuple)
{
	pytype_set_eccentricity(PyTuple_GetItem(tuple, 0));
	pytype_set_semi_major_axis(PyTuple_GetItem(tuple, 1));
}

double Ellipse::focal_parameter()
{
	return semi_major_axis * (1 - eccentricity * eccentricity) / eccentricity;
}

double Ellipse::semi_latus_rectum()
{
	return semi_major_axis * (1 - eccentricity * eccentricity);
}

double Ellipse::linear_eccentricity()
{
	return eccentricity * semi_major_axis;
}

double Ellipse::semi_minor_axis()
{
	return semi_major_axis * sqrt(1 - eccentricity * eccentricity);
}

double Ellipse::aspect_ratio()
{
	return sqrt(1 - eccentricity * eccentricity);
}

double Ellipse::perifocus_distance()
{
	return semi_major_axis * (1 - eccentricity);
}

double Ellipse::apothecure_distance()
{
	return semi_major_axis * (1 + eccentricity);
}

PyObject* Ellipse::pytype_focal_parameter()
{
	return PyFloat_FromDouble(focal_parameter());
}

PyObject* Ellipse::pytype_semi_latus_rectum()
{
	return PyFloat_FromDouble(semi_latus_rectum());
}

PyObject* Ellipse::pytype_linear_eccentricity()
{
	return PyFloat_FromDouble(linear_eccentricity());
}

PyObject* Ellipse::pytype_semi_minor_axis()
{
	return PyFloat_FromDouble(semi_minor_axis());
}

PyObject* Ellipse::pytype_aspect_ratio()
{
	return PyFloat_FromDouble(aspect_ratio());
}

PyObject* Ellipse::pytype_perifocus_distance()
{
	return PyFloat_FromDouble(perifocus_distance());
}

PyObject* Ellipse::pytype_apothecure_distance()
{
	return PyFloat_FromDouble(apothecure_distance());
}